File: update-java/src/main/java/com/mqd/updatejava/download/ApkInstaller.java
package com.mqd.updatejava.download;

import android.content.Context;
import android.content.Intent;
import android.net.Uri;
import android.os.Build;
import android.provider.Settings;



import java.io.File;
import java.io.FileInputStream;
import java.io.FileOutputStream;
import java.io.InputStream;
import java.net.HttpURLConnection;
import java.net.URL;

import androidx.core.content.FileProvider;

/**
 * APK 下载 + 安装器。
 */
public class ApkInstaller {

    private static final int CONNECT_TIMEOUT_MS = 10000;
    private static final int READ_TIMEOUT_MS = 20000;
    private static final int BUFFER_SIZE = 8 * 1024;
    private static final String UPDATE_DIR_NAME = "update";

    private static String fileProviderAuthority = "";

    public static void configure(String authority) {
        fileProviderAuthority = authority;
    }

    public static File updateDir(Context context) {
        File base = context.getExternalCacheDir();
        if (base == null) base = context.getCacheDir();
        return new File(base, UPDATE_DIR_NAME);
    }

    public static String apkFileName(String version) {
        String safe = (version != null && !version.trim().isEmpty()) ? version.trim() : "update";
        safe = safe.replaceAll("[^A-Za-z0-9._-]", "_");
        return safe + ".apk";
    }

    public static File apkFile(Context context, String version) {
        return new File(updateDir(context), apkFileName(version));
    }

    public interface ProgressCallback {
        void onProgress(int percent);
    }

    /**
     * 断点续传下载（同步方法，需在后台线程调用）。
     */
    public static boolean download(String url, File destFile, long expectedSize, ProgressCallback onProgress) {
        File partFile = new File(destFile.getParentFile(), destFile.getName() + ".part");
        HttpURLConnection conn = null;
        try {
            destFile.getParentFile().mkdirs();

            if (isDownloaded(destFile, expectedSize)) {
                if (onProgress != null) onProgress.onProgress(100);
                return true;
            }

            long existing = 0L;
            if (partFile.exists()) {
                existing = partFile.length();
            }
            if (expectedSize > 0 && existing > expectedSize) {
                partFile.delete();
                existing = 0L;
            }
            if (expectedSize > 0 && existing == expectedSize) {
                return finalizePart(partFile, destFile, expectedSize, onProgress);
            }

            conn = (HttpURLConnection) new URL(url).openConnection();
            conn.setRequestMethod("GET");
            conn.setConnectTimeout(CONNECT_TIMEOUT_MS);
            conn.setReadTimeout(READ_TIMEOUT_MS);
            conn.setInstanceFollowRedirects(true);
            if (existing > 0) {
                conn.setRequestProperty("Range", "bytes=" + existing + "-");
            }

            boolean append;
            int code = conn.getResponseCode();
            if (code == HttpURLConnection.HTTP_PARTIAL) {
                append = true;
            } else if (code == HttpURLConnection.HTTP_OK) {
                append = false;
                existing = 0L;
                partFile.delete();
            } else {
                return false;
            }

            long remaining = conn.getContentLength();
            long total = expectedSize > 0 ? expectedSize :
                    (remaining > 0 ? existing + remaining : -1L);
            long downloaded = existing;
            int lastPercent = -1;

            InputStream input = conn.getInputStream();
            FileOutputStream output = new FileOutputStream(partFile, append);
            byte[] buffer = new byte[BUFFER_SIZE];
            int read;
            while ((read = input.read(buffer)) != -1) {
                output.write(buffer, 0, read);
                downloaded += read;
                if (total > 0) {
                    int percent = (int) ((downloaded * 100) / total);
                    if (percent > 100) percent = 100;
                    if (percent != lastPercent && onProgress != null) {
                        lastPercent = percent;
                        onProgress.onProgress(percent);
                    }
                }
            }
            output.flush();
            output.close();
            input.close();

            return finalizePart(partFile, destFile, expectedSize, onProgress);
        } catch (Exception e) {
            return false;
        } finally {
            if (conn != null) conn.disconnect();
        }
    }

    private static boolean finalizePart(File partFile, File destFile, long expectedSize, ProgressCallback onProgress) {
        if (expectedSize > 0 && partFile.length() != expectedSize) {
            partFile.delete();
            return false;
        }
        if (destFile.exists()) destFile.delete();
        if (!partFile.renameTo(destFile)) {
            try {
                FileInputStream fis = new FileInputStream(partFile);
                FileOutputStream fos = new FileOutputStream(destFile);
                byte[] buf = new byte[BUFFER_SIZE];
                int r;
                while ((r = fis.read(buf)) != -1) fos.write(buf, 0, r);
                fis.close();
                fos.close();
                partFile.delete();
            } catch (Exception e) {
                return false;
            }
        }
        if (onProgress != null) onProgress.onProgress(100);
        return true;
    }

    public static boolean isDownloaded(File file, long expectedSize) {
        if (!file.exists() || file.length() <= 0) return false;
        if (expectedSize > 0) return file.length() == expectedSize;
        return true;
    }

    public static void clearOutdatedApks(File dir, String keepFileName) {
        File[] files = dir.listFiles();
        if (files == null) return;
        String keepPart = keepFileName + ".part";
        for (File file : files) {
            if (!file.isFile()) continue;
            String name = file.getName();
            boolean isApk = name.toLowerCase().endsWith(".apk");
            boolean isPart = name.toLowerCase().endsWith(".apk.part");
            if ((isApk || isPart) && !name.equals(keepFileName) && !name.equals(keepPart)) {
                file.delete();
            }
        }
    }

    /**
     * 拉起系统安装器。使用自实现的 FileProvider 提供 content URI。
     */
    public static boolean installApk(Context context, File file) {
        try {
            String authority = !fileProviderAuthority.isEmpty() ?
                    fileProviderAuthority : context.getPackageName() + ".updatejava.fileprovider";
            Uri uri;
            if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.N) {
                uri = FileProvider.getUriForFile(context, authority, file);
            } else {
                uri = Uri.fromFile(file);
            }
            Intent intent = new Intent(Intent.ACTION_VIEW);
            intent.setDataAndType(uri, "application/vnd.android.package-archive");
            intent.addFlags(Intent.FLAG_GRANT_READ_URI_PERMISSION);
            intent.addFlags(Intent.FLAG_ACTIVITY_NEW_TASK);
            context.startActivity(intent);
            return true;
        } catch (Exception e) {
            return false;
        }
    }

    public static boolean canInstall(Context context) {
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O) {
            return context.getPackageManager().canRequestPackageInstalls();
        }
        return true;
    }

    public static void gotoUnknownSourceSetting(Context context) {
        if (Build.VERSION.SDK_INT < Build.VERSION_CODES.O) return;
        try {
            Intent intent = new Intent(Settings.ACTION_MANAGE_UNKNOWN_APP_SOURCES);
            intent.setData(Uri.parse("package:" + context.getPackageName()));
            intent.addFlags(Intent.FLAG_ACTIVITY_NEW_TASK);
            context.startActivity(intent);
        } catch (Exception e) {
            try {
                context.startActivity(new Intent(Settings.ACTION_MANAGE_UNKNOWN_APP_SOURCES)
                        .addFlags(Intent.FLAG_ACTIVITY_NEW_TASK));
            } catch (Exception ignored) {
            }
        }
    }
}

File: update-lib/src/main/java/com/mqd/updatelib/download/ApkInstaller.kt
package com.mqd.updatelib.download

import android.content.Context
import android.content.Intent
import android.net.Uri
import android.os.Build
import android.provider.Settings
import androidx.core.content.FileProvider
import androidx.core.net.toUri
import kotlinx.coroutines.CancellationException
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext
import java.io.File
import java.io.FileOutputStream
import java.net.HttpURLConnection
import java.net.URL

/**
 * apk 下载 + 拉起系统安装器。
 *
 * 下载失败给可读结果（返回 false），不抛异常、不卡死，由上层 toast/通知提示。
 * 下载写入 `<version>.apk.part`，完成且 size 校验通过后才 rename 为正式 `<version>.apk`，
 * 避免坏文件被提升为可安装包。网络中断保留 `.part` 以便断点续传。
 *
 * @author quzhuligpt@gmail.com
 * @since 2026/6/18
 */
object ApkInstaller {

    private const val CONNECT_TIMEOUT_MS = 10000
    private const val READ_TIMEOUT_MS = 20000
    private const val BUFFER_SIZE = 8 * 1024

    private const val UPDATE_DIR_NAME = "update"

    /** FileProvider authority，可通过 configure 设置 */
    private var fileProviderAuthority: String = ""

    /** 配置 FileProvider authority */
    fun configure(authority: String) {
        fileProviderAuthority = authority
    }

    /** 下载目录：优先 externalCacheDir，为 null 时回退 cacheDir。 */
    fun updateDir(context: Context): File {
        val base = context.externalCacheDir ?: context.cacheDir
        return File(base, UPDATE_DIR_NAME)
    }

    /** 按版本号命名，做版本隔离，避免不同版本同名 asset 复用到错误缓存。 */
    fun apkFileName(version: String): String {
        val safe = version.trim().ifBlank { "update" }.replace(Regex("[^A-Za-z0-9._-]"), "_")
        return "$safe.apk"
    }

    fun apkFile(context: Context, version: String): File {
        return File(updateDir(context), apkFileName(version))
    }

    /**
     * 断点续传下载到 [destFile]（实际先写 `destFile.part`），通过 [onProgress] 回调 0..100 进度。
     *
     * - 已有完整正式包（size 命中）→ 直接成功。
     * - `.part` 已存在且 `< expectedSize` → `Range: bytes=N-` 续传；服务器 206 续 / 200 从头覆盖。
     * - 完成后 `length == expectedSize` 才 rename 转正；不匹配删除 + 返回失败。
     * - 网络异常保留 `.part` 以便下次续传。
     */
    suspend fun download(
        url: String,
        destFile: File,
        expectedSize: Long,
        onProgress: (Int) -> Unit
    ): Boolean = withContext(Dispatchers.IO) {
        val partFile = File(destFile.parentFile, destFile.name + ".part")
        var conn: HttpURLConnection? = null
        try {
            destFile.parentFile?.mkdirs()
            // 已有完整正式包：跳过下载
            if (isDownloaded(destFile, expectedSize)) {
                onProgress(100)
                return@withContext true
            }
            var existing = if (partFile.exists()) partFile.length() else 0L
            // .part 大小超出期望（坏残留）→ 删除重下；恰好等于期望的完整 part 留给下面转正
            if (expectedSize > 0 && existing > expectedSize) {
                partFile.delete()
                existing = 0L
            }
            // .part 已完整但尚未转正：直接校验转正，省一次请求
            if (expectedSize > 0 && existing == expectedSize) {
                return@withContext finalizePart(partFile, destFile, expectedSize, onProgress)
            }

            conn = (URL(url).openConnection() as HttpURLConnection).apply {
                requestMethod = "GET"
                connectTimeout = CONNECT_TIMEOUT_MS
                readTimeout = READ_TIMEOUT_MS
                instanceFollowRedirects = true
                if (existing > 0) {
                    setRequestProperty("Range", "bytes=$existing-")
                }
            }
            val append: Boolean
            when (conn.responseCode) {
                HttpURLConnection.HTTP_PARTIAL -> append = true          // 206 续传
                HttpURLConnection.HTTP_OK -> {                           // 200 不支持 Range，从头覆盖
                    append = false
                    existing = 0L
                    if (partFile.exists()) partFile.delete()
                }
                else -> return@withContext false
            }
            val remaining = conn.contentLength.toLong()
            val total = when {
                expectedSize > 0 -> expectedSize
                remaining > 0 -> existing + remaining
                else -> -1L
            }
            var downloaded = existing
            var lastPercent = -1
            conn.inputStream.use { input ->
                FileOutputStream(partFile, append).use { output ->
                    val buffer = ByteArray(BUFFER_SIZE)
                    while (true) {
                        val read = input.read(buffer)
                        if (read == -1) break
                        output.write(buffer, 0, read)
                        downloaded += read
                        if (total > 0) {
                            val percent = ((downloaded * 100) / total).toInt().coerceIn(0, 100)
                            if (percent != lastPercent) {
                                lastPercent = percent
                                onProgress(percent)
                            }
                        }
                    }
                    output.flush()
                }
            }
            finalizePart(partFile, destFile, expectedSize, onProgress)
        } catch (e: CancellationException) {
            throw e
        } catch (e: Exception) {
            // 网络中断保留 .part 以便续传，不删除
            false
        } finally {
            conn?.disconnect()
        }
    }

    /**
     * 完成校验并把 `.part` 转为正式 apk。
     * size 已知则必须匹配，否则删除 `.part` 并返回失败（防坏包转正）。
     */
    private fun finalizePart(
        partFile: File,
        destFile: File,
        expectedSize: Long,
        onProgress: (Int) -> Unit
    ): Boolean {
        if (expectedSize > 0 && partFile.length() != expectedSize) {
            partFile.delete()
            return false
        }
        if (destFile.exists()) destFile.delete()
        if (!partFile.renameTo(destFile)) {
            // rename 失败（跨卷等）兜底复制
            partFile.copyTo(destFile, overwrite = true)
            partFile.delete()
        }
        onProgress(100)
        return true
    }

    /**
     * 缓存命中判断：
     * - 已知大小（expectedSize > 0）：文件存在且大小完全一致，视为已下载。
     * - 未知大小（expectedSize == 0，备用源场景）：文件存在且非空即视为已下载。
     */
    fun isDownloaded(file: File, expectedSize: Long): Boolean {
        if (!file.exists() || file.length() <= 0) return false
        if (expectedSize > 0) return file.length() == expectedSize
        return true
    }

    /**
     * 清理下载目录下除 [keepFileName]（及其 `.part`）外的其它 apk / 残留 part，避免缓存目录堆积。
     */
    fun clearOutdatedApks(dir: File, keepFileName: String) {
        val files = dir.listFiles() ?: return
        val keepPart = "$keepFileName.part"
        for (file in files) {
            if (!file.isFile) continue
            val name = file.name
            val isApk = name.endsWith(".apk", ignoreCase = true)
            val isPart = name.endsWith(".apk.part", ignoreCase = true)
            if ((isApk || isPart) && name != keepFileName && name != keepPart) {
                file.delete()
            }
        }
    }

    /** 拉起系统安装器；启动失败（无安装器等）返回 false，由上层兜底提示。 */
    fun installApk(context: Context, file: File): Boolean {
        return try {
            val authority = fileProviderAuthority.ifEmpty { "${context.packageName}.updatelib.fileprovider" }
            val uri: Uri = if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.N) {
                FileProvider.getUriForFile(context, authority, file)
            } else {
                Uri.fromFile(file)
            }
            val intent = Intent(Intent.ACTION_VIEW).apply {
                setDataAndType(uri, "application/vnd.android.package-archive")
                addFlags(Intent.FLAG_GRANT_READ_URI_PERMISSION)
                addFlags(Intent.FLAG_ACTIVITY_NEW_TASK)
            }
            context.startActivity(intent)
            true
        } catch (e: Exception) {
            false
        }
    }

    fun canInstall(context: Context): Boolean {
        return if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O) {
            context.packageManager.canRequestPackageInstalls()
        } else {
            true
        }
    }

    fun gotoUnknownSourceSetting(context: Context) {
        if (Build.VERSION.SDK_INT < Build.VERSION_CODES.O) return
        try {
            val intent = Intent(Settings.ACTION_MANAGE_UNKNOWN_APP_SOURCES).apply {
                data = "package:${context.packageName}".toUri()
                addFlags(Intent.FLAG_ACTIVITY_NEW_TASK)
            }
            context.startActivity(intent)
        } catch (e: Exception) {
            try {
                context.startActivity(
                    Intent(Settings.ACTION_MANAGE_UNKNOWN_APP_SOURCES)
                        .addFlags(Intent.FLAG_ACTIVITY_NEW_TASK)
                )
            } catch (ignored: Exception) {
            }
        }
    }
}
